src/rust/engine/process_execution/src/nailgun/mod.rs

```diff
@@ -248,28 +260,30 @@ impl CapturedWorkdir for CommandRunner {
     // Streams to read child output from
     let (stdio_write, stdio_read) = child_channel::<ChildOutput>();
     let (_stdin_write, stdin_read) = child_channel::<ChildInput>();
+    let client_req = construct_nailgun_client_request(
+      req2,
+      client_main_class,
+      client_args,
+      client_workdir.clone(),
+    )?;
 
-    let nails_command = self
-      .async_semaphore
-      .with_acquired(move || {
-        // Get the port of a running nailgun server (or a new nailgun server if it doesn't exist)
-        nailgun_pool.connect(
-          nailgun_name.clone(),
-          nailgun_req,
-          workdir_for_this_nailgun1,
-          nailgun_req_digest,
-          build_id,
-          store,
-          req.input_files,
-          workunit_store,
-        )
-      })
+    // Get the port of a running nailgun server (or a new nailgun server if it doesn't exist)
+    let nails_command = nailgun_pool
+      .connect(
+        nailgun_name.clone(),
+        nailgun_req,
+        workdir_for_this_nailgun1,
+        nailgun_req_digest,
+        build_id,
+        store,
+        req.input_files,
+        workunit_store,
+      )
       .map_err(|e| format!("Failed to connect to nailgun! {}", e))
       .inspect(move |_| debug!("Connected to nailgun instance {}", &nailgun_name3))
-      .and_then(move |nailgun_port| {
+      .and_then(move |(nailgun_port, nailgun_guard)| {
```

Comment:
If I understand the code correctly, this will introduce the contstraint that only one client can be connected to an entry in the map at once. Is that intentional? The previous code didn't impose that constraint (locked the entire map until `connect` ended, and then unlocked it).

I can believe it would be necessary, but I can't think of a good reason why.

Reply:
Previously there was also a lock around individual processes, which was acquired to get the port. The lock was being dropped at the end of the connect routine. The problem with this is that a server could be restarted (due to new fingerprint) while a client was still running on the old version. This way is safer, a thread has to wait until the last client finishes before restarting a nailgun. I have a branch that uses a RwLock around the NailgunProcess as we discussed, it is much faster, but it has some concurrency/dependency bug right now that I am trying to work out. Another possible direction would be to have multiple servers running with the same fingerprint, and parallelize that way, which would mean keeping the constraint of one client running per server.

Reply:
> The problem with this is that a server could be restarted (due to new fingerprint) while a client was still running on the old version

As it stands now in the Pants model, the fingerprints are one-per-task, so if we are restarting a server while a client is still running, it means that two tasks are trying to use the same nailgun, which is a bug in a different place.
In general I agree that this is safer, but given the Pants model and how it wants to use these, we can assume that no server will attempt to restart while a client is running. (I think I had @illicitonion check my math on that once, but couldn't hurt to get his eyes and maybe @stuhood 's on it).

Reply:
If that is the case I'm unclear why we have a specific case to restart the nailgun with a new fingerprint, if it changes. Should we raise an error instead if what you say above is true?

Reply:
I agree that a map-entry-specific RwMutex is probably a reasonable answer, by the way :)

Reply:
Okay, thanks for the explanation, I will remove the hold on the mutex guard.

>I agree that a map-entry-specific RwMutex is probably a reasonable answer, by the way :)

I have a follow up PR that does this. I am running into some issues with the JVM compile execution graph that might be solved if we don't have to keep the ReadLock alive by passing it through a futures tree. However I would like to keep the scope of this PR to fixing the OOM, and then focus on improving parallelism in the follow up.

Reply:
There is some very difficult to debug socket error that happens when we let multiple clients run on a nailgun concurrently. It manifests a socket error in the nailgun stdout and a hanging task that timesout in the pants client. I am opting for keeping the guard alive for the length of the run, which effectively limits 1:1 client server relationship. This seems like something we want to do anyway moving forward with non nailMain tools. For concurrency in the future we will want out nailgun process pool to support multiple nailguns per tool name.

Reply:
> Talking to @stuhood this seems like something we want to do anyway moving forward with non nailMain tools.

Yeah, that was the "while we are in beta" plan, so I'm not super against holding the guard for now _iff_  there is a very lengthy and explicit TODO for that, possibly linking to this conversation and the issue about supporting multiple tools per name.